File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrDocStateUIExtension.cs
using System.Threading.Tasks;
using Tessa.Extensions.Default.Shared;
using Tessa.Platform.Storage;
using Tessa.UI.Cards;

namespace Tessa.Extensions.Default.Client.UI
{
    /// <summary>
    /// Расширение на обновление виртуальной карточки состояния документа со стороны клиента.
    /// Обновление можно быть по кнопке "Обновить" или при обновлении после успешного сохранения.
    /// </summary>
    public sealed class KrDocStateUIExtension :
        CardUIExtension
    {
        #region Base Overrides

        public override async Task Reopening(ICardUIExtensionContext context)
        {
            ICardEditorModel editor = context.UIContext.CardEditor;
            if (editor == null || context.Model == null || context.GetRequest == null)
            {
                return;
            }

            int? stateID = null;
            if (editor.CurrentOperationType == CardEditorOperationType.SaveAndRefresh)
            {
                // идентификатор этой карточки можно менять, поэтому актуальный идентификатор при рефреше после сохранения будет в той карточке,
                // которая сохранялась (и была успешно сохранена, раз запущен рефреш)
                stateID = context.Model.Card.Sections["KrDocStateVirtual"].RawFields.Get<int?>("StateID");
            }

            if (!stateID.HasValue)
            {
                // либо это обновление без сохранения, либо по какой-то причине поле было пустым
                stateID = context.Model.Card.TryGetInfo()?.TryGet<int?>(DefaultExtensionHelper.StateIDKey);
            }

            if (!stateID.HasValue)
            {
                return;
            }

            context.GetRequest.Info[DefaultExtensionHelper.StateIDKey] = stateID.Value;
        }

        #endregion
    }
}

File: WebClient SDK/src/default/ui/krDocStateUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext, CardEditorOperationType } from 'tessa/ui/cards';
import { tryGetFromInfo } from 'tessa/ui';
import { DotNetType, createTypedField } from 'tessa/platform';

/**
 * Расширение на обновление виртуальной карточки состояния документа со стороны клиента.
 */
export class KrDocStateUIExtension extends CardUIExtension {

  public shouldExecute(context: ICardUIExtensionContext) {
    return context.card.typeId === 'e83a230a-f5fc-445e-9b44-7d0140ee69f6'; // KrDocStateTypeID
  }

  public reopening(context: ICardUIExtensionContext) {
    const editor = context.uiContext.cardEditor;
    if (editor == null
      || context.model == null
      || context.getRequest == null
    ) {
        return;
    }

    let stateId: number | null = null;
    if (editor.currentOperationType === CardEditorOperationType.SaveAndRefresh) {
      // идентификатор этой карточки можно менять, поэтому актуальный идентификатор при рефреше после сохранения будет в той карточке,
      // которая сохранялась (и была успешно сохранена, раз запущен рефреш)
      stateId = context.model.card.sections.get('KrDocStateVirtual')!.fields.get('StateID');
    }

    if (stateId == null) {
      // либо это обновление без сохранения, либо по какой-то причине поле было пустым
      stateId = tryGetFromInfo<number | null>(context.model.card.tryGetInfo() || {}, 'StateID');
    }

    if (stateId == null) {
      return;
    }

    context.getRequest.info['StateID'] = createTypedField(stateId, DotNetType.Int32);
  }

}